File: treee/224-basic-calculator.scala
// Memory Limit Exceeded

object Solution {
    def calculate(s: String): Int = {
        import scala.collection.mutable.ListBuffer
        def next(s: String): Option[(Either[Int, Char], String)] = {
            if (s.isEmpty) {
                None
            } else if (Character.isSpaceChar(s.head)) {
                next(s.tail)
            } else if (Character.isDigit(s.head)) {
                var t = ""
                var s1 = s
                while (!s1.isEmpty && Character.isDigit(s1.head)) {
                    t = t :+ s1.head
                    s1 = s1.tail
                }
                Some(Left(Integer.parseInt(t)), s1)
            } else {
                Some(Right(s.head), s.tail)
            }
        }

        def iterNonRec(s: String): ListBuffer[Either[Int, Char]] = {
            var l = ListBuffer[Either[Int, Char]]()
            var res = next(s)
            while (res.isDefined) {
                val q = res.get._2
                l = l :+ res.get._1
                res = next(q)
            }
            l
        }

        def iter(s: String): Seq[Either[Int, Char]] = {
            next(s) match {
                case None => Nil
                case Some((t, q)) => {
                    t +: iter(q)
                }
            }
        }

        def calcOne(a: Int, b: Int, op: Char): Int = {
            op match {
                case '+' => a + b
                case '-' => b - a
            }
        }

        def calc(seq: ListBuffer[Either[Int, Char]], stkNum: Seq[Int], stkOp: Seq[Char]): Int = {
            if (seq.isEmpty) {
                if (stkNum.isEmpty) {
                    0
                } else {
                    stkNum.head
                }
            } else {
                seq.head match {
                    case Left(n) => {
                        if (stkOp.isEmpty || stkOp.head == '(') {
                            calc(seq.tail, n +: stkNum, stkOp)
                        } else {
                            val r = calcOne(n, stkNum.head, stkOp.head)
                            calc(seq.tail, r +: stkNum.tail, stkOp.tail)
                        }
                    }
                    case Right(op) => {
                        if (op == ')') {
                            if (!stkOp.tail.isEmpty) {
                                val r = calcOne(stkNum.head, stkNum.tail.head, stkOp.tail.head)
                                calc(seq.tail, r +: stkNum.tail.tail, stkOp.tail.tail)
                            } else {
                                calc(seq.tail, stkNum, stkOp.tail)
                            }
                        } else {
                            calc(seq.tail, stkNum, op +: stkOp)
                        }
                    }
                }
            }
        }

        calc(iterNonRec(s), Seq(), Seq())
    }
}

File: treee/224-basic-calculator.cpp
class Solution {
public:
    
    pair<int, char> next(string&s, int *idx, int *f) {
        while (*idx < s.length() && s[*idx] == ' ')
            ++(*idx);
        if (*idx == s.length()) {
            *f = 1;
            return make_pair(0, ' ');
        }
        if (isdigit(s[*idx])) {
            string t = "";
            while (isdigit(s[*idx]))
                t += s[(*idx)++];
            return make_pair(atoi(t.c_str()), ' ');
        }
        return make_pair(0, s[(*idx)++]);
    }
    
    int calc(int a, int b, char op) {
        if (op == '+')
            return b + a;
        return b - a;
    }
    
    int calculate(string s) {
        int i = 0;
        stack<int> stkNum;
        stack<char> stkOp;
        int f = 0;
        while (i < s.length()) {
            auto t = next(s, &i, &f);
            if (f)
                break;
            if (t.second != ' ') {
                if (t.second == ')') {
                    stkOp.pop();
                    if (!stkOp.empty()) {
                        int r = stkNum.top();
                        stkNum.pop();
                        r = calc(r, stkNum.top(), stkOp.top());
                        stkOp.pop();
                        stkNum.pop();
                        stkNum.push(r);
                    }
                }
                else {
                    stkOp.push(t.second);
                }
            }
            else {
                if (stkOp.empty() || stkOp.top() == '(')
                    stkNum.push(t.first);
                else {
                    int r = calc(t.first, stkNum.top(), stkOp.top());
                    stkNum.pop();
                    stkNum.push(r);
                    stkOp.pop();
                }
            }
        }
        
        return stkNum.top();
    }
};
